### PR TITLE
refactor: simplified constraints and type mapping

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR finalizes the constraint and type mapping behavior.

### CODE REVIEW

File: src/helpers/ConstrainHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  }
+  constrainedModel.type = getTypeFromMapping(typeMapping, {constrainedModel, renderer: context.renderer, propertyKey: context.propertyKey});
+  return constrainedModel;
+}
+
+export function constrainMetaModel<R extends AbstractRenderer>(typeMapping: TypeMapping<R>, constrainRules: Constraints, context: ConstrainContext<R, MetaModel>): ConstrainedMetaModel {
+  const constrainedName = constrainRules.modelName({modelName: context.metaModel.name});
+  const newContext = {...context, constrainedName};
+  if (newContext.metaModel instanceof ObjectModel) {
+    return constrainObjectModel(typeMapping, constrainRules, {...newContext, metaModel: newContext.metaModel});
```
Comment (magicmatatjahu) [inline src/helpers/ConstrainHelpers.ts:182]: You can simplify it:

```ts
const newContext = {...context, constrainedName, metaModel: context.metaModel};

....

return constrainReferenceModel(typeMapping, constrainRules, newContext);
```

**Review by magicmatatjahu — APPROVED** [review]
Awesome, one suggestion, but you don't have to propagate it :)

File: src/helpers/ConstrainHelpers.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/helpers/ConstrainHelpers.ts:182]: I tried, TS is throwing errors that the types for metaModel, when done like that, would be `MetaModel`, which does not match with `ReferenceModel` (for example).

Alternatively, I could do:
```
const newContext = {...context, constrainedName};
....
return constrainReferenceModel(typeMapping, constrainRules, newContext as any);
```

But that seems forced as well, and kinda shady 🤔 What you think?

Comment (magicmatatjahu) [reply src/helpers/ConstrainHelpers.ts:182]: Ok, so please don't propagate my suggestion :) LGTM!

Comment (jonaslagoni) [reply src/helpers/ConstrainHelpers.ts:182]: Alright 👍

**Review by magicmatatjahu — APPROVED** [review]

**Pull request merged**